rfcs/20220223-convert_the_store_from_vuex_to_pinia.md

```diff
@@ -0,0 +1,127 @@
+
+- [ ] TBD
+  
+- [ ] TBD
+  
+
+**Milestone:** https://github.com/WordPress/openverse-frontend/milestone/5
+  
+
+## Rationale
+
+Currently we use Vuex for global state management. This works fine for Vue 2 but has some considerable disadvantages for Vue 3, especially when it comes to TypeScript support and composition-api usage. We should switch to Pinia instead. It is the new [official default recommendation]( https://blog.vuejs.org/posts/vue-3-as-the-new-default.html) of the Vue project for state management instead of Vuex. 
+
+Pinia also has better interoperability between Vue 2's composition-api and Vue 3, and would allow us to decouple our state-management API usage from Nuxt. Currently we rely on Nuxt's home-grown `useStore` composable from the `@nuxtjs/composition-api` extension on top of `@vue/composition-api` but there's no guarantee that that composable will stick around in Nuxt 3. Besides, it is not easy to watch for changes in the store from within the `setup` function, at least when rendering on the server, and it caused problems with rendering the correct search input text on SSR.
+
+Moving to Pinia will allow us to reduce the current verbosity of the store introduced by module, action and mutation constants. In the long run, it will simplify the unit tests setup.
+
+## Alternatives
+
+We could just stick with the current Vuex approach and during the Nuxt bridge upgrade replace all usages of `useStore` with `useNuxtApp().$store` instead. This would introduce more code changes during the Nuxt bridge upgrade BUT would prevent us from having to re-write our entire state management strategy before migrating to Nuxt 3. 
+
+## Potential risks
+- Pinia is currently maintained by a Vue core member, but it's still not the official Vue Vuex implementation. However, its development is closely linked with the development of the new version of Vuex and they strive for the same API and some feature parity, so we should be able to move to Vuex 5 whenever it is released if we decide to do so.
+- Store changes affect the app as a whole, and changes can cause a lot of disruption. To mitigate this, we can ensure that each store rewrite is accompanied with a full rewritten unit test, and try to use types as much as possible.
+
+## Conversion to Pinia
+There are some important differences between Vuex (the version we are currently using)and Pinia in store structure.
+
+### Dividing the store: Vuex modules and Pinia stores
+
+Vuex uses a single store which can have multiple modules that can be namespaced. In our setup we have several namespaced modules inside `src/store` that Nuxt uses to create a single store.
+In Pinia, the equivalent of a Vuex namespaced module is a single store. Namespacing is achieved by giving a store its `id`. All Pinia stores are separate modules, and are usually kept in a `src/stores` folder to emphasize that. To use a store `X` from store `Y`, you need to call `useXStore` inside `storeY`.
```

Comment:
If two stores invoke each other, can this lead to a cyclic import in the JS files?

Reply:
That's a nice point we should add:
> If two or more stores use each other, they cannot create an infinite loop through getters or actions.
But reading each other's state directly in their setup function is not possible, we should use a computed to access other store's state: https://pinia.vuejs.org/cookbook/composing-stores.html

Reply:
In our case, we should probably do `useOtherStore` in the media store, as it uses a lot of other stores. And if media store state or actions are needed in another store, we should use `useMediaStore` inside a computed or an action (which does not create a circular dependency).